lib/inky.ex

```diff
@@ -3,60 +3,232 @@ defmodule Inky do
     @moduledoc false
     @enforce_keys [:display, :hal_state]
     defstruct type: nil,
               hal_state: nil,
               display: nil,
+              wait_type: :nowait,
               pixels: %{}
   end
 
   require Integer
+  require Logger
 
   alias Inky.Commands
   alias Inky.Displays.Display
   alias Inky.PixelUtil
 
+  @push_timeout 5000
+
   @color_map_black %{black: 0, miss: 1}
   @color_map_accent %{red: 1, yellow: 1, accent: 1, miss: 0}
 
+  #
   # API
+  #
 
   @doc """
-  Initializes a display of the given configuration. This function will do some
-  of the necessary preparation to prepare communication with the display.
+  Start a GenServer that deals with the HAL state (initialization of and communication with the display) and pushing pixels to the physical display. This function will do some of the necessary preparation to prepare communication with the display.
 
   ## Parameters
 
     - type: Atom for either :phat or :what
     - accent: Accent color, the color the display supports aside form white and black. Atom, :black, :red or :yellow.
   """
-  def init(type, accent)
-      when type in [:phat, :what] and accent in [:black, :red, :yellow] do
+  def start_link(args \\ %{}) do
+    opts = if(args[:name], do: [name: args[:name]], else: [])
+    GenServer.start_link(__MODULE__, args, opts)
+  end
+
+  @doc """
+  `set_pixels(pid | name, pixels | painter, opts \\ %{push: :await})` set
+  pixels and draw to display (or not!), with data or a painter function.
+
+  set_pixels updates the internal state either with specific pixels or by
+  calling `painter.(x,y,w,h,current_pixels)` for all points in the screen, in
+  an undefined order.
+
+  Currently, the only option checked is `:push`, which represents the minimum
+  pixel pushing policy the caller wishes to apply for their request. Valid
+  values are listed and explained below.
+
+  NOTE: the internal state of Inky will still be updated, regardless of which
+  pushing policy is employed.
+
+  - `:await`: Busy wait until you can push to display, clearing any previously
+    set timeout. This is the default.
+  - `:once`: Push to the display if it is not busy, otherwise, report that it
+    was busy. Only `:await` timeouts are reset if a `:once` push has failed.
+  - `{:timeout, :await}`: Use genserver timeouts to avoid multiple updates.
+    When the timeout triggers, await device with a busy wait and then push to
+    the display. If the timeout previously was :once, it is replaced.
+  - `{:timeout, :once}`: Use genserver timeouts to avoid multiple updates. When
+    the timeout triggers, update the display if not busy. Does not downgrade a
+    previously set `:await` timeout.
+  - `:skip`: Do not push to display. If there has been a timeout previously
+    set, but that has yet to fire, it will remain set.
+  """
+  def set_pixels(pid, arg, opts \\ %{push: :await}),
+    do: GenServer.call(pid, {:set_pixels, arg, opts}, :infinity)
+
+  # TODO: rename to push?
```

Comment:
I think show is the better name. Possibly `update` if `show` feels to definite with all the push options.

Reply:
show is either sync or async to the client, the implementation always awaits. I left it as show to keep things moving along, though.